### ISSUE TITLE
COMPASS: User specified number of cores for a test case

### ISSUE THREAD

**Issue by xylar** [MPAS-Dev/MPAS-Model#581]
We are starting to add various steps to COMPASS ocean test cases (other than calls to `ocean_model`) that require multiple cores (multiple MPI tasks).

At the same time, work is in progress to run test cases from regression suites in parallel, and this requires accurate knowledge of how many cores a given test case needs, which is currently based only on calls to the model, not on other steps.

Currently, changing the number of MPI tasks is an unnecessarily tedious process, particularly when PIO has been configured with a specific number of MPI task in mind in the namelist file.

It would be really great if we had a way to specify (and for a user to alter) the number of MPI tasks once and for all, and then steps within a test case would refer to this one source to determine how many cores to use, how to configure PIO, etc.  

Such an approach would likely require splitting steps that require significantly different numbers of cores into multiple test cases.  For example, for large meshes, mesh creation and generating an initial condition would likely be separate from spin-up runs as we already do for `global_ocean` test cases with large meshes.

**Comment by xylar** [comment]
@matthewhoffman, @mark-petersen, @vanroekel, @pwolfram, I'd appreciate hearing your thoughts.

**Comment by xylar** [comment]
One thought I have is adding a python config file at the level of the test case (e.g. `ocean/global_ocean/QU240/init`) as opposed to the level of the individual steps (e.g. `ocean/global_ocean/QU240/init/e3sm_coupling`).  This config file would be passed to the scripts for each step's `run.py` and would be used to substitute placeholders in the python script that would come from the `config_*.xml`.  I haven't completely thought through how this might work.

**Comment by matthewhoffman** [comment]
I wonder if it would be worth investigating some of the task-parallelism approached that are out there.  I worry this level of complexity is pushing up against the limitations of the handwritten COMPASS system we have.  I forget the name of the task-based framework that @philipwjones has been working with, but I also came across this, which is python based, developed in part at Argonne, and uses Cori as an application example:
https://parsl-project.org/

**Comment by xylar** [comment]
@matthewhoffman, I have discussed parsl with @vanroekel and @jamilgafur.  I don't think parsl would supersede the need for some way to know how many cores a given test case needs and having an easy way to modify this.  But I do fully agree that it would be a good option for parallelizing COMPASS.  It's kind of overkill for the relatively simple case we have, and ultimately it's @vanroekel and @jamilgafur's project so I'll leave the specifics up to them.

My most pressing need isn't to do with parallelizing COMPASS but rather simply running a single test case on different systems where different numbers of cores are available/appropriate.

**Comment by philipwjones** [comment]
The task-parallel framework we're using in CANGA is FleCSI (C++) and probably a little heavyweight for what you're doing here.

FWIW, I have worked with Henry to put together the capability for calling Metis internally as a library on startup rather than requiring a specific part.NN files for each task count in the ocean_model runs.  That would let ocean_model partition itself at runtime by only reading the main graph.info adjacency file and no longer need a separate partition file. We have been able to replicate the existing partitions and it works pretty fast even for high-res. Haven't integrated it into the framework yet because I need to figure out how to support Adrian's new multi-domain approach for sea-ice partitioning. I realize this discussion isn't focused on the ocean_model piece, but just in case it's of interest...
